process_scheduling_algorithms/non_preemptive_priority_scheduling.c

```diff
@@ -0,0 +1,400 @@
+    }
+    avg = sum / n;
+    return avg;
+}
+/**
+ * @brief To calculate the average waiting time of all the processes
+ * @param root pointer to the head of the queue
+ * @returns float average waiting time
+ */
+float CalculateWT(struct node **root)
+{
+    float avg, sum = 0;
+    int n = LLength(root);
+    struct node *ptr = *root;
+    // calculate the completion if not already calculated
+    if (ptr->CT == 0)
+    {
+        CalculateCT(root);
+    }
+    // calculate the total waiting time of all the processes
+    while (ptr != NULL)
+    {
+        ptr->WT = (ptr->TAT - ptr->BT);
+        sum += ptr->WT;
+        ptr = ptr->next;
+    }
+    avg = sum / n;
+    return avg;
+}
+
+// self implemented test case
+void test()
+{
+    // Entered processes
+    printf("ID Priority Arrival Time Burst Time \n");
+    printf("1 0 5 1 \n");
+    printf("2 1 4 2 \n");
+    printf("3 2 3 3 \n");
+    printf("4 3 2 4 \n");
+    printf("5 4 1 5 \n");
+
+    struct node *root = NULL;
+    insert(&root, 1, 0, 5, 1);
+    insert(&root, 2, 1, 4, 2);
+    insert(&root, 3, 2, 3, 3);
+    insert(&root, 4, 3, 2, 4);
+    insert(&root, 5, 4, 1, 5);
+    printf("Average Completion Time is : %f \n", CalculateCT(&root));
+    printf("Average Turn Around Time is : %f \n", CalculateTAT(&root));
+    printf("Average Waiting Time is : %f \n", CalculateWT(&root));
+}
+
+void main()
+{
+    test();
+
+    // For Menu Driven Program
+    /*
+    struct node *root = NULL;
+    int option = 0, prior, bt, id, at;
+    while (option != 7)
+    {
+        printf("1. Enter Process Details \n");
+        printf("2. Delete Process \n");
+        printf("3. Display Process \n");
+        printf("4. Calculate Avg Completion Time \n");
+        printf("5. Calculate Avg Turn Around Time \n");
+        printf("6. Calculate Avg Waiting Time \n");
+        printf("7. Exit. \n");
+        scanf("%d", &option);
+        switch (option)
+        {
+        case 1:
+            printf("Enter Process Id \n");
+            scanf("%d", &id);
+            printf("Enter Process Priority \n");
+            scanf("%d", &prior);
+            printf("Enter Process Arrival Time \n");
+            scanf("%d", &at);
+            printf("Enter Process Burst Time \n");
+            scanf("%d", &bt);
+            insert(&root, id, at, bt, prior);
+            break;
+        case 2:
+            printf("Enter Process Id \n");
+            scanf("%d", &id);
+            delete (&root, id);
+            break;
+        case 3:
+            showList(root);
+            break;
+        case 4:
+            printf("Average Completion Time is : %f \n", CalculateCT(&root));
+            break;
+        case 5:
+            printf("Average Turn Around Time is : %f \n",
+            CalculateTAT(&root)); break;
+        case 6:
+            printf("Average Waiting Time is : %f \n", CalculateWT(&root));
+            break;
+        default:
+            break;
+        }
+    }
+    */
```

Comment:
Good practice that is very used in C/C++ programs.
```suggestion
    */
    return 0;
```